language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaLocateSyntaxTreeUtil.java

```diff
@@ -0,0 +1,91 @@
+package org.ballerinalang.langserver.extensions.ballerina.document;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.SyntaxTree;
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.TextDocument;
+import io.ballerina.tools.text.TextRange;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.eclipse.lsp4j.Range;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This is the BallerinaLocateSyntaxTreeUtil class for related utils in retrieving the location of syntax nodes
+ * for a given selection.
+ */
+public class BallerinaLocateSyntaxTreeUtil {
+    private static final String POSITION = "position";
+    private static final String LOCATE_PATH_PROPERTY = "isNodePath";
+
+    public static JsonElement mapNodePath(Range range, SyntaxTree syntaxTree, JsonElement syntaxTreeJson) {
+        syntaxTreeJson.getAsJsonObject().addProperty(LOCATE_PATH_PROPERTY, true);
+        Node node = CommonUtil.findNode(range, syntaxTree);
+
+        if (node.kind() == SyntaxKind.LIST) {
+            node = node.parent();
+        }
+        if (node == syntaxTree.rootNode()) {
+            return syntaxTreeJson;
+        }
+        return findNodePath(syntaxTreeJson, syntaxTree, node);
+    }
+
+    private static JsonElement findNodePath (JsonElement syntaxTreeJson, SyntaxTree syntaxTree, Node node) {
+        Optional<JsonElement> temp = Optional.of(syntaxTreeJson);
+
+        while (temp.isPresent()) {
+            temp = findChildren(syntaxTree, node, temp.get());
+        }
+        return syntaxTreeJson;
+    }
+
+    private static Optional<JsonElement> findChildren(SyntaxTree syntaxTree, Node node, JsonElement jsonNode) {
+        for (Map.Entry<String, JsonElement> childEntry: jsonNode.getAsJsonObject().entrySet()) {
+            if (childEntry.getValue().isJsonArray()) {
+                for (JsonElement childProps: childEntry.getValue().getAsJsonArray()) {
+                    Integer isPath = evaluateRange(syntaxTree, node, childProps,
```

Comment:
Shall we use boolean flag here?

Reply:
We are checking for 3 different conditions, in order to break the loop at the right time.

Reply:
Cool, that makes sense. Shall we add some comments here, since the flag is an integer it is bit fuzzy to grasp the logic at once. Or, we can use an Enum to represent the state as well, which will be self explaining the logic.

Reply:
Noted. will look into it and make the changes.